buildSrc/src/main/kotlin/io/opentelemetry/gradle/OtelVersionClassPlugin.kt

```diff
@@ -0,0 +1,74 @@
+package io.opentelemetry.gradle
+
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.plugins.JavaPlugin
+import org.gradle.api.plugins.JavaPluginExtension
+import org.gradle.kotlin.dsl.the
+import java.io.File
+
+/**
+ * This gradle plugin will define a new task called generateOtelVersionClass.
+ * This task generates a Java source file that contains the project version
+ * as a string constant. The "compileJava" task is updated to depend on
+ * generateOtelVersionClass, and the project source set is updated to
+ * include the new file.
+ */
+class OtelVersionClassPlugin : Plugin<Project> {
+  override fun apply(project: Project) {
+    project.plugins.apply(JavaPlugin::class.java)
+
+    project.task("generateOtelVersionClass") {
+      doLast {
+        writeFile(project)
+      }
+    }
+    addTaskDependency(project)
+    addSourceDependency(project)
+  }
+
+  private fun writeFile(project: Project) {
+    val group = "${project.group}".replace('.', '/')
+    val projectName = project.name.replace('-', '/')
+    val outDir = buildOutDir(project)
+    val filename = "$group/$projectName/OtelVersion.java"
+    val outFile = File(outDir, filename)
+    val packageName = "${project.group}.${project.name.replace('-', '.')}"
+    val classBody = getClassBody("${project.version}", packageName)
+
+    outFile.parentFile.mkdirs()
+    outFile.writeText(classBody)
+  }
+
+  private fun getClassBody(version: String, packageName: String): String {
+    return """
+      package $packageName;
+      
+      /**
+       * Autogenerated class do not edit.
+       */
+       public final class OtelVersion {
```

Comment:
My only concern with this approach: will devs have to run a gradle build in order to be able to compile in IDEA? Or is IDEA smart enough to run what needs to be run and get this file into the generated source when the project is imported?

Reply:
Yeah, I'm sure a gradle build step has to happen before `Resource` can compile. Isn't that the same with autovalue tho?

Another option was to do string substitution on the actual source tree, but that leaves uncommitted changes. Open to alternatives.

Reply:
Here is one idea of how to compromise: https://gist.github.com/breedx-splk/01278cc3eb94415353cac1f9cdce89c7

The idea is instead of referencing a static field constant, try to do the same with reflection, and if that fails, fall back to the previous behavior. I don't love it.